### PR TITLE
[plugin-json] Add custom matcher for json arrays validation

### LINKED ISSUE #3643: Add possibility to check that one JSON array contains another JSON array

### PR DESCRIPTION
**Pull request by EDbarvinsky**
Closes #3643

- [x] Implementation
- [x] Unit tests
- [x] Integration tests
- [ ] Docs

[depends on](https://github.com/lukas-krecan/JsonUnit/pull/617)

### CODE REVIEW

File: vividus-plugin-json/src/test/java/org/vividus/json/steps/JsonArrayContainsAnyOfTests.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    void shouldThrowExceptionParameterNotArray()
+    {
+        performValidationTest(() -> matcher.setParameter(JSON_OBJECT_NODE.toString()));
+    }
+
+    @Test
+    void shouldThrowExceptionActualNotArray()
+    {
+        String parameter = "[1]";
+        matcher.setParameter(parameter);
```
Comment (EDbarvinsky) [inline vividus-plugin-json/src/test/java/org/vividus/json/steps/JsonArrayContainsAnyOfTests.java:97]: remove this 2 lines

Comment (EDbarvinsky) [reply vividus-plugin-json/src/test/java/org/vividus/json/steps/JsonArrayContainsAnyOfTests.java:97]: and compose test to the previous one

File: vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java [hunk 2]
```diff
@@ -74,16 +75,19 @@
     private final JsonUtils jsonUtils;
     private final ISoftAssert softAssert;
     private final IAttachmentPublisher attachmentPublisher;
+    private final Map<String, Matcher<Object>> customJsonMatchers;
 
     public JsonSteps(FluentEnumConverter fluentEnumConverter, JsonContext jsonContext, VariableContext variableContext,
-            JsonUtils jsonUtils, ISoftAssert softAssert, IAttachmentPublisher attachmentPublisher)
+            JsonUtils jsonUtils, ISoftAssert softAssert, IAttachmentPublisher attachmentPublisher,
+                     Map<String, Matcher<Object>> customJsonMatchers)
```
Comment (EDbarvinsky) [inline vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java:82]: or replace with setters approach?

File: vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonArrayContainsAnyOf.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    public boolean matches(Object actual)
+    {
+        List<JsonNode> actualNodes = convert(jsonUtils.toJson(actual));
+        return actualNodes.stream().anyMatch(expectedNodes::contains);
+    }
+
+    @Override
+    public void describeTo(Description description)
+    {
+        description.appendText("json does not contains any elements from ").appendValue(expectedNodes);
```
Comment (valfirst) [inline vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonArrayContainsAnyOf.java:57]: how does final error message look like?

File: vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java (hunk 2 above)
Comment (valfirst) [reply vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java:82]: nope, injecting via constructor is an appropriate way

File: vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java [hunk 4]
```diff
@@ -74,16 +75,19 @@
     private final JsonUtils jsonUtils;
     private final ISoftAssert softAssert;
     private final IAttachmentPublisher attachmentPublisher;
+    private final Map<String, Matcher<Object>> customJsonMatchers;
```
Comment (valfirst) [inline vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java:78]:
```suggestion
    private final Map<String, ParametrizedMatcher<Object>> customJsonMatchers;
```
?

File: vividus-plugin-json/src/main/resources/vividus-plugin/spring.xml [hunk 5]
```diff
@@ -12,10 +12,19 @@
 
     <bean id="json-context" class="org.vividus.json.variable.JsonContextVariable" />
 
+    <util:map id="jsonMatchers" map-class="java.util.HashMap">
+        <entry key="anyOf">
+            <bean class="org.vividus.json.steps.JsonArrayContainsAnyOf">
+                <constructor-arg ref="jsonUtils-LowerCamelCase" />
+            </bean>
+        </entry>
+    </util:map>
```
Comment (valfirst) [inline vividus-plugin-json/src/main/resources/vividus-plugin/spring.xml:21]: it's allowed to declare named beans, no need to create Map, Spring will do it

File: vividus-plugin-json/src/test/java/org/vividus/json/steps/JsonPatchStepsTests.java [hunk 6]
```diff
@@ -42,7 +40,7 @@
     private VariableContext variableContext;
 
     @Test
-    void patchJsonFile() throws JsonProcessingException
```
Comment (valfirst) [inline vividus-plugin-json/src/test/java/org/vividus/json/steps/JsonPatchStepsTests.java:45]: noisy change

File: vividus-plugin-json/src/test/java/org/vividus/json/steps/JsonStepsTests.java [hunk 7]
```diff
@@ -145,7 +146,7 @@ void beforeEach()
     {
         JsonPathUtils.setJacksonConfiguration();
         steps = new JsonSteps(new FluentEnumConverter(), jsonContext, variableContext, new JsonUtils(), softAssert,
-                attachmentPublisher);
+                attachmentPublisher, Collections.emptyMap());
```
Comment (valfirst) [inline vividus-plugin-json/src/test/java/org/vividus/json/steps/JsonStepsTests.java:149]:
```suggestion
                attachmentPublisher, Map.of());
```

File: vividus-tests/src/main/resources/story/integration/JSON.story [hunk 8]
```diff
@@ -227,3 +227,14 @@ Then JSON `
   "required": [ "productId", "productName", "price", "tags" ]
 }
 `
+
+Scenario: Validate anyOf json-unit matcher
+Then JSON element from `${json}` by JSON path `$.store` is equal to `
+{
+  "book":"#{json-unit.matches:anyOf}[{\"category\":\"poetry\",\"author\":\"Yakub Kolas\",\"title\":\"Novaya zyamlya\",\"isbn\":null,\"price\":8.95},{\"category\": \"fiction\",\"author\": \"Evelyn Waugh\",\"title\": \"Sword of Honour\",\"price\": 12.99,\"hardcover\": false}]",
```
Comment (valfirst) [inline vividus-tests/src/main/resources/story/integration/JSON.story:234]: please format in a readable way

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonArrayContainsAnyOf.java (hunk 3 above)
Comment (EDbarvinsky) [reply vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonArrayContainsAnyOf.java:57]: `json does not contains any elements from <[1, 2]>`

File: vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonArrayContainsAnyOf.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    private List<JsonNode> convert(String jsonArray)
+    {
+        JsonNode node = jsonUtils.readTree(jsonArray);
+        Validate.isTrue(node.isArray(), "Expected type `ARRAY`, but found `%s`: %s", node.getNodeType(),
+            node.toPrettyString());
+        List<JsonNode> result = new ArrayList<>();
+        node.elements().forEachRemaining(result::add);
+        return result;
```
Comment (uarlouski) [inline vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonArrayContainsAnyOf.java:67]: can't we return node.elements()?

**Pull request closed without merging**